---
title: "Benchmarking computational efficiency of TreeSE methods"
author: "Giulio Benedetti and Leo Lahti"
date: "`r Sys.Date()`"
---


```{r setup, echo = FALSE}
# no need to show all code, let us just show the benchmarking results
# code is in the Rmd file for those who want to dig in deeper.
# call packages to load data sets
knitr::opts_chunk$set(message = FALSE, warning = FALSE)

library(ggplot2)                # plotting
theme_set(theme_bw(20))

library(dplyr)                  # pipe operator
library(mia)                    # data sets and analysis
library(microbiomeDataSets)     # data sets
library(curatedMetagenomicData) # data sets
library(scater)                 # beta diversity
```

# Computational efficiency of TreeSE methods

This document provides some benchmarking results comparing the
computational efficiency in standard operations between
`TreeSummarizedExperiment` (tse) and `phyloseq` (pseq) containers for
microbiome data.

## Introduction

To estimate the time efficiency of homologous tasks but with different
data structures (`tse` or `pseq`), we have benchmarked a set of
standard data manipulation routines on data sets of varying numbers of
samples and features, as reported below.


## Data sets

The comparisons are based on publicly available data sets from the
curatedMetagenomicData project. See the homepage for the details and
references. 


```{r placeholders}
# define data sets
data_sets <- c("AsnicarF_2017", "GlobalPatterns", "VincentC_2016", "BackhedF_2015", "SilvermanAGutData", "ZeeviD_2015", "SongQAData", "SprockettTHData", "GrieneisenTSData")

# set seed and define sample size
set.seed(3)
sample_size <- 10

# assign working variables with a placeholder to work with them inside the for loop.
len_set <- length(data_sets)
tse <- TreeSummarizedExperiment()
tmp <- list()
```

```{r dataframe, echo = FALSE}
# create NA data frame to store the execution times throughout the for loop
df <- data.frame(Features = rep(NA, len_set),
                 Samples = rep(NA, len_set),
		 Melt = rep(NA, len_set),
		 Transform = rep(NA, len_set),
		 Agglomerate = rep(NA, len_set),
		 AlphaEstimation = rep(NA, len_set),
		 BetaEstimation = rep(NA, len_set),
		 row.names = data_sets)

df <- rbind(df[rep(1:len_set, 2), ]) %>%
        mutate(ObjectType = c(rep("tse", len_set), rep("pseq", len_set)),
	       MeltCommand = c(rep("mia_meltAssay", len_set), rep("phyloseq_psMelt", len_set)),
	       TransformCommand = c(rep("mia_transformSamples", len_set), rep("microbiome_transform", len_set)),
	       AgglomerateCommand = c(rep("mia_agglomerateByRank", len_set), rep("phyloseq_tax_glom", len_set)),
	       AlphaCommand = c(rep("mia_estimateDiversity", len_set), rep("microbiome_diversity", len_set)),
	       BetaCommand = c(rep("runMDS", len_set), rep("ordinate", len_set)),
	       AssayValues = rep(c("relative_abundance", "counts", "relative_abundance", "relative_abundance", "counts", "relative_abundance", "counts", "counts", "counts"), 2))
```

Execution times for different experiments, data sets and containers are evaluated with a recursive approach. Results are stored into `df`.

```{r conditions, echo = FALSE}
# conditions for the for loop (calculating conditions outside of the for loop improves efficiency)
condition_1 <- data_sets == "GlobalPatterns"
condition_2 <- data_sets %in% c("SilvermanAGutData", "SongQAData", "SprockettTHData", "GrieneisenTSData")
condition_3 <- data_sets == "GrieneisenTSData"
condition_4 <- data_sets %in% c("AsnicarF_2017", "VincentC_2016", "BackhedF_2015", "ZeeviD_2015")
condition_5 <- !(data_sets %in% c("SilvermanAGutData", "GrieneisenTSData"))
```

# Compare execution times

```{r iteration, echo = FALSE}
for (data_set in data_sets) {

  # define index of current data set
  cur_set <- which(data_sets == data_set)

  # load mia
  if (condition_1[cur_set]) {

    mapply(data, list = data_set, package = "mia")
    tse <- eval(parse(text = data_set))
    mapply(data, list = data_set, package = "phyloseq")
    pseq <- eval(parse(text = data_set))

    # load microbiomeDataSets
  } else if (condition_2[cur_set]) {

    tse <- eval((parse(text = paste0(data_set, "()"))))

    if (condition_3[cur_set]) {
      rowData(tse) <- DataFrame(lapply(rowData(tse), unfactor))
    }

    pseq <- makePhyloseqFromTreeSummarizedExperiment(tse)

    # load curatedMetagenomicData
  } else if (condition_4[cur_set]) {

    tmp <- curatedMetagenomicData(paste0(data_set, ".relative_abundance"), dryrun = FALSE, counts = TRUE)

    tse <- tmp[[1]]

    pseq <- makePhyloseqFromTreeSummarizedExperiment(tse,
                                                     abund_values = df$AssayValues[cur_set])

  }
  
  # select big data sets at least as large as sample_size
  if (dim(tse)[2] >= sample_size) {
  
    # random subsetting
    subset_names <- sample(colnames(tse), sample_size)
    tse <- tse[ , colData(tse)@rownames %in% subset_names]

    # store dimensions
    df$Features[cur_set] <- dim(tse)[1]
    df$Features[cur_set + len_set] <- dim(tse)[1]
    df$Samples[cur_set] <- dim(tse)[2]
    df$Samples[cur_set + len_set] <- dim(tse)[2]

    # test melting for tse
    start.time1 <- Sys.time()
    molten_tse <- mia::meltAssay(tse,
                       abund_values = df$AssayValues[cur_set],
                       add_row_data = TRUE,
                       add_col_data = TRUE)
    end.time1 <- Sys.time()
    df$Melt[cur_set] <- end.time1 - start.time1

    # test melting for pseq
    start.time2 <- Sys.time()
    molten_pseq <- phyloseq::psmelt(pseq)
    end.time2 <- Sys.time()
    df$Melt[cur_set + len_set] <- end.time2 - start.time2

    # test transformation for tse
    start.time1 <- Sys.time()
    trans_tse <- mia::transformSamples(tse,
                                  method = "log10",
                                  pseudocount = 1,
                                  abund_values = df$AssayValues[cur_set])
    end.time1 <- Sys.time()
    df$Transform[cur_set] <- end.time1 - start.time1

    # test transformation for pseq
    start.time2 <- Sys.time()
    trans_pseq <- microbiome::transform(pseq,
                                    transform = "log10p",
                                    target = "sample")
    end.time2 <- Sys.time()
    df$Transform[cur_set + len_set] <- end.time2 - start.time2

    # test agglomeration for tse
    start.time1 <- Sys.time()
    tse_phylum <- agglomerateByRank(tse,
                               rank = "Phylum",
                               na.rm = TRUE)
    end.time1 <- Sys.time()
    df$Agglomerate[cur_set] <- end.time1 - start.time1

    # test agglomeration for pseq
    start.time2 <- Sys.time()
    pseq_phylum <- phyloseq::tax_glom(pseq,
                                  taxrank = "Phylum")
    # na.rm = TRUE by default in tax_glom
    end.time2 <- Sys.time()
    df$Agglomerate[cur_set + len_set] <- end.time2 - start.time2

    # test alpha diversity estimation for tse
    start.time1 <- Sys.time()
    alpha_tse <- mia::estimateDiversity(tse,
                             abund_values = df$AssayValues[cur_set],
                             index = "shannon",
                             name = "shannon")
    end.time1 <- Sys.time()
    df$AlphaEstimation[cur_set] <- end.time1 - start.time1

    # test alpha diversity estimation for pseq
    start.time2 <- Sys.time()
    alpha_pseq <- microbiome::diversity(pseq,
                     index = "shannon")
    end.time2 <- Sys.time()
    df$AlphaEstimation[cur_set + len_set] <- end.time2 - start.time2

    if (condition_5[cur_set]) {

      # test beta diversity estimation for tse
      start.time1 <- Sys.time()
      beta_tse <- scater::runMDS(tse, FUN = vegan::vegdist, name = "MDS_BC", exprs_values = df$AssayValues[cur_set])
      pcoa_tse <- scater::plotReducedDim(beta_tse, "MDS_BC")
      end.time1 <- Sys.time()
      df$BetaEstimation[cur_set] <- end.time1 - start.time1

      # test beta diversity estimation for pseq
      start.time2 <- Sys.time()
      beta_pseq <- phyloseq::ordinate(pseq, "MDS", "bray")
      pcoa_pseq <- phyloseq::plot_ordination(pseq, beta_pseq, type = "samples")
      end.time2 <- Sys.time()
      df$BetaEstimation[cur_set + len_set] <- end.time2 - start.time2

    }
    
  }

}

rm("condition_1", "condition_2", "condition_3", "condition_4", "condition_5")
```

### Melting

Melting versus samples.

```{r melting_samples}
ggplot(df, aes(x = Samples, y = Melt, color = MeltCommand, shape = as.factor(Features))) +
  geom_point() +
  labs(title = "Melting comparison", x = "Number of samples", y = "Execution time in s", color = "method:", shape = "number of features:", caption = "Execution time of melting as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

Melting versus features.

```{r melting_features}
ggplot(df, aes(x = Features, y = Melt, color = MeltCommand, shape = as.factor(Samples))) +
  geom_point() +
  labs(title = "Melting comparison",
       x = "Number of features",
       y = "Execution time in s",
       color = "method:", shape = "number of samples:",
       caption = "Execution time of melting as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```

### Transformations

```{r transformation_samples}
ggplot(df, aes(x = Samples, y = Transform, color = TransformCommand, shape = as.factor(Features))) +
  geom_point() +
  labs(title = "Transformation comparison",
       x = "Number of samples",
       y = "Execution time in s",
       color = "method:",
       shape = "number of features:",
       caption = "Execution time of transformation as a function of number of samples") +
  theme(legend.position = "bottom")
```

```{r transformation_features}
ggplot(df, aes(x = Features, y = Transform, color = TransformCommand, shape = as.factor(Samples))) +
  geom_point() +
  labs(title = "Transformation comparison",
       x = "Number of features",
       y = "Execution time in s",
       color = "method:", shape = "number of samples:",
       caption = "Execution time of transformation as a function of number of features") +
  theme(legend.position = "bottom")
```

### Agglomeration

Agglomeration versus samples

```{r agglomeration_samples}
ggplot(df, aes(x = Samples, y = Agglomerate, color = AgglomerateCommand, shape = as.factor(Features))) +
  geom_point() +
  labs(title = "Agglomeration comparison",
       x = "Number of samples",
       y = "Execution time in s",
       color = "method:",
       shape = "number of features:",
       caption = "Execution time of agglomeration as a function of number of samples") +
  theme(legend.position = "bottom")
```

Agglomeration versus features

```{r agglomeration_features}
ggplot(df, aes(x = Features, y = Agglomerate, color = AgglomerateCommand, shape = as.factor(Samples))) +
  geom_point() +
  labs(title = "Agglomeration comparison",
       x = "Number of features",
       y = "Execution time in s",
       color = "method:",
       shape = "number of samples:",
       caption = "Execution time of agglomeration as a function of number of features") +
  theme(legend.position = "bottom")
```

### Alpha diversity 


Beta diversity versus sample size.

```{r alpha_samples}
ggplot(df, aes(x = Samples, y = AlphaEstimation, color = AlphaCommand, shape = as.factor(Features))) +
  geom_point() +
  labs(title = "Alpha Diversity Estimation comparison",
       x = "Number of samples",
       y = "Execution time in s",
       color = "method:",
       shape = "number of features:",
       caption = "Execution time of estimating alpha diversity as a function of number of samples") +
  theme(legend.position = "bottom")
```

Beta diversity versus feature size.

```{r alpha_features}
ggplot(df, aes(x = Features, y = AlphaEstimation, color = AlphaCommand, shape = as.factor(Samples))) +
  geom_point() +
  labs(title = "Alpha Diversity Estimation comparison",
       x = "Number of features",
       y = "Execution time in s",
       color = "method:",
       shape = "number of samples:",
       caption = "Execution time of estimating alpha diversity as a function of number of features") +
  theme(legend.position = "bottom")
```




### Beta diversity 


Beta diversity versus sample size.

```{r beta_samples}
ggplot(df, aes(x = Samples, y = BetaEstimation, color = BetaCommand, shape = as.factor(Features))) +
  geom_point() +
  labs(title = "Beta Diversity Estimation comparison",
       x = "Number of samples",
       y = "Execution time in s",
       color = "method:",
       shape = "number of features:",
       caption = "Execution time of estimating beta diversity as a function of number of samples") +
  theme(legend.position = "bottom")
```


Beta diversity versus feature size.

```{r beta_features}
ggplot(df, aes(x = Features, y = BetaEstimation, color = BetaCommand, shape = as.factor(Samples))) +
  geom_point() +
  labs(title = "Beta Diversity Estimation comparison",
       x = "Number of features",
       y = "Execution time in s",
       color = "method:",
       shape = "number of samples:",
       caption = "Execution time of estimating beta diversity as a function of number of features") +
  theme(legend.position = "bottom")
```



# Mean difference in execution time

```{r meantime}
mean_time <- df %>% group_by(ObjectType) %>%
                    summarize(mean_melt = mean(Melt),
		              mean_transform = mean(Transform),
			      mean_agglomerate = mean(Agglomerate),
			      mean_alpha = mean(AlphaEstimation),
			      mean_beta = mean(BetaEstimation, na.rm = TRUE))
mean_time
```

