---
title: "Benchmarking computational efficiency of TreeSE methods"
author: "Giulio Benedetti and Leo Lahti"
date: "`r Sys.Date()`"
---

```{r include = FALSE}
source("data.R", local = knitr::knit_global())
```

```{r dataframe, echo = FALSE}
# create NA data frame to store the execution times throughout the for loop
df <- data.frame(Features = rep(NA, len_set),
                 Samples = rep(NA, len_set),
		 Melt = rep(NA, len_set),
		 row.names = data_sets)

df <- rbind(df[rep(1:len_set, 2), ]) %>%
        mutate(ObjectType = c(rep("tse", len_set), rep("pseq", len_set)),
	       MeltCommand = c(rep("mia_meltAssay", len_set), rep("phyloseq_psMelt", len_set)),
	       AssayValues = rep(c("relative_abundance", "counts", "relative_abundance", "counts", "counts", "counts", "counts"), 2))
```

```{r iteration, echo = FALSE}
for (data_set in data_sets) {

  # define index of current data set
  cur_set <- which(data_sets == data_set)

  # load mia
  if (condition_1[cur_set]) {

    mapply(data, list = data_set, package = "mia")
    tse <- eval(parse(text = data_set))
    mapply(data, list = data_set, package = "phyloseq")
    pseq <- eval(parse(text = data_set))

    # load microbiomeDataSets
  } else if (condition_2[cur_set]) {

    tse <- eval((parse(text = paste0(data_set, "()"))))

    if (condition_3[cur_set]) {
      rowData(tse) <- DataFrame(lapply(rowData(tse), unfactor))
    }

    pseq <- makePhyloseqFromTreeSummarizedExperiment(tse)

    # load curatedMetagenomicData
  } else if (condition_4[cur_set]) {

    tmp <- curatedMetagenomicData(paste0(data_set, ".relative_abundance"), dryrun = FALSE, counts = TRUE)

    tse <- tmp[[1]]

    pseq <- makePhyloseqFromTreeSummarizedExperiment(tse,
                                                     abund_values = df$AssayValues[cur_set])

  }
  
  # select big data sets at least as large as sample_size
  if (dim(tse)[2] >= sample_size) {
  
    # random subsetting
    subset_names <- sample(colnames(tse), sample_size)
    tse <- tse[ , colData(tse)@rownames %in% subset_names]

    # store dimensions
    df$Features[cur_set] <- dim(tse)[1]
    df$Features[cur_set + len_set] <- dim(tse)[1]
    df$Samples[cur_set] <- dim(tse)[2]
    df$Samples[cur_set + len_set] <- dim(tse)[2]

    # test melting for tse
    start.time1 <- Sys.time()
    molten_tse <- mia::meltAssay(tse,
                       abund_values = df$AssayValues[cur_set],
                       add_row_data = TRUE,
                       add_col_data = TRUE)
    end.time1 <- Sys.time()
    df$Melt[cur_set] <- end.time1 - start.time1

    # test melting for pseq
    start.time2 <- Sys.time()
    molten_pseq <- phyloseq::psmelt(pseq)
    end.time2 <- Sys.time()
    df$Melt[cur_set + len_set] <- end.time2 - start.time2
  
  }
  
}

rm("condition_1", "condition_2", "condition_3", "condition_4", "condition_5")
```

```{r melting_features}
ggplot(df, aes(x = Features, y = Melt, color = MeltCommand, shape = as.factor(Samples))) +
  geom_point() +
  labs(title = "Melting comparison",
       x = "Number of features",
       y = "Execution time in s",
       color = "method:", shape = "number of samples:",
       caption = "Execution time of melting as a function of number of features") +
  theme(legend.position = "bottom")
```